dftimewolf/lib/collectors/gsheets.py

```diff
@@ -0,0 +1,257 @@
+        state, name=name, critical=critical)
+    self._credentials = None
+    self._spreadsheet_id = ''
+    self._sheets_names: List[str] = []
+    # These are mandatory columns required by Timesketch.
+    self._mandatory_columns = ['message', 'datetime', 'timestamp_desc']
+    self._all_sheets = False
+    self._validate_columns = True
+
+  # pylint: disable=arguments-differ
+  def SetUp(self,
+            spreadsheet: str,
+            sheets_names: List[str],
+            validate_columns: bool = True) -> None:
+    """Sets up a a Google Sheets collector.
+
+    Args:
+      spreadsheet: ID or URL of the sheet to pull data from
+      sheets_names: List of sheets names inside the spreadsheet to parse. 'All'
+        will parse all sheets inside a spreadsheet.
+      validate_columns: Check if mandatory columns required by Timesketch is
+        present in the sheets.
+    """
+    self._credentials = self._GetCredentials()
+    self._spreadsheet_id = self._ValidateSpreadSheetId(spreadsheet)
+    self._sheets_names = sheets_names
+    if 'all' in (sheet.lower() for sheet in sheets_names):
+      self._all_sheets = True
+    self._validate_columns = validate_columns
+
+  def Process(self) -> None:
+    """Copies entries from Google Sheets."""
+
+    try:
+      # Retrive list of sheets in the spreadsheet
+      service = self._BuildSheetsResource(self._credentials)
+      # Pylint can't see the spreadsheets method.
+      # pylint: disable=no-member
+      result = service.spreadsheets().get(
+          spreadsheetId=self._spreadsheet_id).execute()
+      spreadsheet_title = result.get('properties', {}).get('title')
+      sheets = (result.get('sheets', []))
+
+      for sheet in sheets:
+        if not sheet.get('properties'):
+          continue
+
+        sheet_title = sheet.get('properties', {}).get('title')
+
+        if not self._all_sheets and sheet_title not in self._sheets_names:
+          continue
+
+        self.logger.info('Parsing sheet: {0:s}'.format(sheet_title))
+
+        df = self._ExtractEntiresFromSheet(self._spreadsheet_id, sheet_title)
+
+        if df is None or df.empty:
+          continue
+
+        output_file = tempfile.NamedTemporaryFile(
+            mode='w', delete=False, encoding='utf-8', suffix='.csv')
+        output_path = output_file.name
+        self.logger.info(
+            'Downloading results to of sheet "{0:s}" to {1:s}'.format(
+                sheet_title, output_path))
+
+        df.to_csv(index=False, na_rep='NaN', path_or_buf=output_file)
+
+        self.logger.success(
+            'Downloaded results of sheet "{0:s}" to {1:s}'.format(
+                sheet_title, output_path))
+        output_file.close()
+        sheet_csv_file = containers.File(
+            name='{0:s}_{1:s}'.format(spreadsheet_title, sheet_title),
+            path=output_path)
+        self.state.StoreContainer(sheet_csv_file)
+
+    except (RefreshError, DefaultCredentialsError) as exception:
```

Comment:
No, this is different.

Reply:
Even if it is covered in the preflight, it is still good to handle it here. Keeping things modular.

Reply:
I mean that we should still check for the exception. I am not reimplementing anything, just adding an extra check and not assuming that this exception won't happen because another module will handle it. 

This is also the same in the workspace_audit module (  https://github.com/log2timeline/dftimewolf/blob/49bb9bf0225e2d9e3c2c2a494a08cf93f6517c38/dftimewolf/lib/collectors/workspace_audit.py#L169 )

Reply:
My 2cts on this is that the reason we have these preflights so that we don't have to do all these checks in separate modules, that's why I would choose to leave this out here. @tomchop WDYT?

Reply:
Another factor to take into account is whether the preflight is added to the recipe. If not, then this exception would need to be caught. I don't think it's super clear-cut, and it's not a hill I'm willing to die on; I'm happy to leave the check here and revisit this paradigm if we feel like it's getting too messy.

Reply:
GCPTokenCheck requires a project name. I don't have one in this case. I will mark this as resolved unless if there is another way to add this preflight-check in.